### ISSUE TITLE
Memory stays high after dropping torrent

### ISSUE THREAD

**Issue by Fesaa** [anacrolix/torrent#930]
Heya!

Once a torrent is finished downloading (There is a check every 5m, that drops a torrent is torrent.Complete is true.), I drop it with torrent.Drop(), and rename it from the map I use to keep track of them. However, my application's memory usage stays up, not going down unless I restart. 
After a few downloads, it's at ~400mb, while if I hadn't downloaded anything it stabilizes around 25mb.

I've tried looking at a heapdump with pprof, but I understand far too little of it to figure out what might be going wrong with it.

Am I supposed to remove a torrent on the client somewhere, or close something else first? Any help would be super welcome ❤️ 

Thanks for the amazing project!
~ Amelia

**Comment by anacrolix** [comment]
Sending or linking your heap dump would be the easiest. Another option is if you can create a unit test that adds your torrents, gets the memory use up then does the Drop demonstrating the memory use (it'll be enough if you can get it to this stage for me to debug it). Thanks!

**Comment by Fesaa** [comment]
Heya,

For my actual application; made a heap on startup, while downloading two torrents, and then ±2m after they were both done. [20240419_mp_heaps.zip](https://github.com/anacrolix/torrent/files/15036998/20240419_mp_heaps.zip)

I've also written a quick program that shows the same behaviour, but stripped down the logic a lot. Run without arguments for a little web server, with the exact same dropping logic as explained above. (`curl http://127.0.0.1:8080/torrent` to start them). Run with any argument, for a simple add torrents and client.DownloadAll(). 
[20240419_torrent_930.zip](https://github.com/anacrolix/torrent/files/15037014/20240419_torrent_930.zip)

Can change the torrents it downloads in `main.go:15-18`, just used two iso's, as these download relatively fast.
For the example with the web server, I had 25MB startup, 60-70MB during download and 50-60MB constantly after. 

Thanks for looking into this!

**Comment by anacrolix** [comment]
@Fesaa when you run your demo program, how are you extracting memory profiles?

**Comment by anacrolix** [comment]
I've added envpprof to the demo (tests/issue-930). I also found gops which might be another way to do this without adding my instrumentation package.

I removed a large allocation that occurs at init which improved memory use in general. I suspect something is holding a `*Torrent` but I'm not able to work out what yet. The Go tooling in this area is lacking. Memory use does decline a few minutes after dropping, I suspect either GC or goroutines holding objects are expiring (like peer connections, DHT etc.).

**Comment by anacrolix** [comment]
I also noticed that go-libutp creates large buffers (~1MB/conn), and they don't really shrink back down. I don't think this is a leak, after you run some torrents, you will continue to receive incoming utp connections. The buffers can be made smaller, but I remember raising them to increase throughput.

**Comment by Fesaa** [comment]
I read them from activity monitor (macOS) and Monitor (Gnome). I didn't add any memory profiles there sorry, but I normally just add net/http/pprof and start a net/http server in a goroutine at startup. 

For my actual application I read them from docker stats, although I'm starting to suspect docker stats is reporting way more memory than it's actually using, as the sum of these is higher than what the host OS is reporting to be used. While this is obviously out of scope here, it should make the issue a lot less big than I initially described.

**Comment by anacrolix** [comment]
You should notice a significant improvement with the fix in master, but I don't think this issue is resolved yet (unless that satisfies you).

**Comment by Fesaa** [comment]
I'll try and find time some time in the following days to test this out. Thanks!

**Comment by Fesaa** [comment]
Heya! 

I'm seeing 10MB less usage on startup, and idle until a download starts. Running the test program above on my host machine, gets me to 100-110MB, and then goes down again slowly. 
It's currently at 45MB here, but still going down. While the same program running in docker goes up to ~200MB, and just stays there. 
The container is running Ubuntu, with the same behavior in alpine. And the host is arch, but that shouldn't cause this much of a difference. And is not going down in the same time span. 

![image](https://github.com/anacrolix/torrent/assets/77553571/db406ce8-b51d-449f-88f3-9a3a7a442232)

Assuming docker is just reporting the memory usage wrong, or reading something else. I'm fine with this, it's for a personal application and restarting it every so often is no issue whatsoever for me.

**Comment by anacrolix** [comment]
I suspect Docker here, what version of Go are you using? There were reports in older versions of not being able to reclaim memory easily, perhaps you're running into this.

Let's close this since there's nothing else actionable here for now. We don't know if there's actually any more leaks in anacrolix/torrent, but definitely reopen it if we find something.

**Comment by istancescu** [comment]
I've observed a similar behaviour while spinning multiple torrent magnets at the same time, but downloading only the metadata and dropping the torrent. The memory went up to 200 mb, but never reclaimed or reclaimed just a fair bit (like few mbs)

This was happening both in a docker container or just in a builded binary, and this is the most similar issue to the one I have.

Perhaps there still might be a memory leak somewhere.

**Comment by anacrolix** [comment]
Could you provide a pprof heap profile? inuse_space is the one I need @istancescu. Similarly with @Fesaa if you are still interested.

**Comment by Fesaa** [comment]
Hey, I've got my application running and exporting metrics to my grafana instance. Is `inuse_bytes` the correct one? I haven't run torrents in a while, but can try and see if I have the same behaviour as mentioned above if it is. 

<img width="364" alt="Image" src="https://github.com/user-attachments/assets/83068667-d037-4855-a3fa-25cc67a59cb2" />

I haven't had any issues with memory as far as I know, and don't have the time for more at the moment. Sorry

**Comment by anacrolix** [comment]
https://pkg.go.dev/net/http/pprof#hdr-Usage_examples

**Comment by istancescu** [comment]
Attaching here my heap dump, let me know if this is sufficient.

[pprof.___go_build_plex_organizer_go.alloc_objects.alloc_space.inuse_objects.inuse_space.002.pb.gz](https://github.com/user-attachments/files/19725498/pprof.___go_build_plex_organizer_go.alloc_objects.alloc_space.inuse_objects.inuse_space.002.pb.gz)

**Comment by anacrolix** [comment]
Thanks, I'll check it in a bit but that file name looks right.

**Comment by anacrolix** [comment]
Bummer. I'll need a goroutine listing as well.

**Comment by istancescu** [comment]
[goroutine.txt](https://github.com/user-attachments/files/19906872/goroutine.txt)

Hope this helps

**Comment by anacrolix** [comment]
So you have a torrent.Client still, but you dropped the Torrent? Looks like all the goroutines for the Torrent have stopped, but the heap profile shows the data is still around anyway. My best guess is there are pieces stuck in the request strategy ordering or something holding all the Torrent's data structures alive. Let me know if that sounds right.

**Comment by istancescu** [comment]
Yes, I do keep the client alive, but I am dropping the torrents. I've also observed if I use torrents with huge size  or big number of files the memory usage expands more quickly.

**Comment by anacrolix** [comment]
Please include a heap profile when you have an excess of memory in use. I will spin that out as a separate issue if confirmed, but I do know there's some low hanging fruit for improving memory use in a few areas relating to piece/file tracking but I need samples to confirm.

**Comment by anacrolix** [comment]
Just to emphasis I can't progress on this without heap profiles.

**Comment by istancescu** [comment]
I know, I'll get back soon with few more data, I had a long week so far.

**Comment by anacrolix** [comment]
No worries. Incidentally I stumbled on a 9 year something old inefficiency in some locks that meant several data types were occupying 300-600 bytes each more than necessary. Particularly Piece. That's fixed on master.

**Comment by istancescu** [comment]
Hello again, 

This is with go: 1.24.2

`github.com/anacrolix/torrent v1.58.1`

I didn't get the chance to test the new version yet, I'll come with that too back later.

I've had almost 30 sequential torrent metadata calls that lead up to around ~163mb of memory usage at peak.

![Image](https://github.com/user-attachments/assets/e7942bcc-842e-4804-a2c3-0e819822d0ad)

After some time, it recovers but doesn't go below 70 mb of usage.

![Image](https://github.com/user-attachments/assets/17a287f4-d179-439f-b9f5-7879358647e6)

this is what the pprof debug window shows, mostly a lot of memory remain stucks in makePieces, perhaps with the new version that will get minimized.

![Image](https://github.com/user-attachments/assets/75733a7c-69a7-4c66-bcda-198a978aa1e5)

[goroutines.txt](https://github.com/user-attachments/files/20169899/goroutines.txt)

[heap.pb.gz](https://github.com/user-attachments/files/20170001/heap.pb.gz)

**Comment by istancescu** [comment]
I know the GC is not instant, but after this memory rises it seems not to be recovering even after an hour (I didn't let the app run for more than that). The file sizes are 1-10, and for above samples metadata was from 100KB to 700KB per torrent (20-30 torrents).

The memory increase is much slower for smaller torrents with lower metadata size (2kb), but it happens, although it's hard to get that many samples for it to have a significant memory usage right now.

I'll let this run for a couple more hours and I'll come back with an edit to see if the current run recovers more memory in time or stays at the current value (151MB 23.14 EEST).
